spanset: assert that batches don't access store local and unreplicated RangeID local keys

iskettaneh opened the pull request:
This PR adds the following test-only assertions:
1) Generated batches don't touch store-local keys.
2) Generated batches don't touch unreplicated RangeID local keys.

We disable the check in exactly 3 locations we know that we currently
touch those keys.

Fixes: #156537

Release note: None

iskettaneh committed:
kvserver/spanset: introduce forbidden spans matchers

This commit introduces the concept of forbidden spans, which will
allow us to assert our batches don't modify Raft's engine keys.

iskettaneh committed:
spanset: assert that batches don't access store local and unreplicated RangeID local keys

This commit adds the following test-only assertions:
1) Generated batches don't touch store-local keys.
2) Generated batches don't touch unreplicated RangeID local keys.

We disable the check in exactly 2 locations we know that we currently
touch those keys.

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,132 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	}
+
+	if span.EndKey == nil {
+		span.EndKey = span.Key.Next()
```
This could allocate twice. Not a huge deal in tests, but still: how about doing the comparisons statelessly? Could we use the existing `overlaps` func? Also, we can probably avoid `Prevish`. 

For the `{Key: nil, EndKey: x}` case, essentially, the bad case is `start < x <= end` (as opposed to `start <= x < end` for `{Key: x, EndKey: nil}`).

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,132 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+
+		if modifiedSpan.Overlaps(getRangeIDLocalSpan(startRangeID)) {
+			return errors.Errorf("overlapping an unreplicated rangeID span")
+		}
```
It seems like `DecodeRangeIDKey` gives us all we need. The `infix` bit is equal to [localRangeIDUnreplicatedInfix](https://github.com/cockroachdb/cockroach/blob/2b759bf2e7d059bdb53caf40a626ad00fa8bc85e/pkg/keys/constants.go#L111) when we're in the unreplicated space. But it's a bit annoying that, due to the `(x-eps,x]` semantics of our span, we should be checking `start < x <= end` here, which means `"u" < infix+suffix+detail <= "v"`.

But we know that `infix+suffix+detail` can't equal `u` or `v`, because `suffix` must't be empty (`DecodeRangeIDKey` would return an error).

So... looks like it's sufficient to check that `infix == "u"`?

iskettaneh committed:
spanset: Add Overlaps() helper function

Analogous to how we have a spanset helper contains() that
understands the special span representation: [x-eps,x).
This commit adds Overlaps that expects the same span
representation.

Moreover, this commit makes this special span representation
explicit by introducing a new type called `TrickySpan`.

pav-kv commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -83,8 +83,14 @@ type Span struct {
+	// explicitly.
+	forbiddenSpansMatchers []func(roachpb.Span) error
+	allowUndeclared        bool
+	allowForbidden         bool
```
Pre-existing, but made stronger in this PR: it seems that this type shouldn't mix the job of representing a set of spans and that of [test-only] verifying it / forbidding keys. I would make it the responsibility of a wrapper (either existing one like `spanSetBatch`, or make a new one).

Might need a prereq PR that does that (and as a drive-by clean up all the weirdly- or unused stuff in `SpanSet`).

iskettaneh replied:
Will do it in another PR
<!-- Reviewable comment -OeNcN23436FK_3YtzrM:b-zegkzz -->
<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,108 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+		return errors.Errorf("overlaps with store local keys")
+	}
+
+	return nil
```
Does the order matter here, or you made the `Overlaps` aware of both cases? Could be also:

```golang
if spanset.Overlaps(span, roachpb.Span{
	Key:    keys.LocalStorePrefix,
	EndKey: keys.LocalStoreMax,
}) {
	return error
}
return nil
```

iskettaneh replied:
Now I made sure that ordering matters, and that the ordering is correct.
<!-- Reviewable comment -OeNu7cN5-4CYG-ibcr6:b-5hfvmo -->
<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,108 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	// local keys.
+	if spanset.Contains(span, fullRangeIDLocalSpans) {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
```
`Contains` can return true if `A == B`, right? But we're specifically interested whether there are parts of `A` that are inside `B`, as well as outside.

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,108 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	fullRangeIDLocalSpans := roachpb.Span{
+		Key:    keys.MakeRangeIDUnreplicatedPrefix(0),
+		EndKey: keys.LocalRangePrefix,
+	}
```
After getting inside `Local/RangeID`, I think we can afford a simplified check which does not depend much on the nil-ness of keys.

Inside `Local/RangeID`, we know that all keys *must* have an encoded RangeID.

TODO: finish the comment.

iskettaneh replied:
I followed your recommendation, and it made things a lot simpler!
<!-- Reviewable comment -OeO0qdM7TqWWf6x6XEg:brydkxx -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,108 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	if span.EndKey != nil {
+		endRangeID, err := keys.DecodeRangeIDPrefix(span.EndKey)
+		if err != nil {
+			panic(err) // cannot happen
```
Done.
<!-- Reviewable comment -OeO1Uy82G00AVBPDBql:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/batcheval/cmd_truncate_log.go:
```diff
@@ -118,7 +118,8 @@ func TruncateLog(
 	// raft.
 	// We can pass zero as nowNanos because we're only interested in SysBytes.
-	ms, err := storage.ComputeStats(ctx, readWriter, start, end, 0 /* nowNanos */)
+	ms, err := storage.ComputeStats(ctx,
```
Done.
<!-- Reviewable comment -OeO1imd1i2Kt2gI_qrK:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -382,11 +382,20 @@ func (s *SpanSet) checkAllowed(
+		}
+		return s2.Key.IsPrev(s1.EndKey) &&
+			(s2.EndKey == nil || s1.EndKey.Compare(s2.EndKey) == 0)
+	}
```
Done. You were right, the changes I made was to make these helper functions accept the TrickySpan either as the first or second parameter. I followed your suggestion with the TrickySpan.
<!-- Reviewable comment -OeNsvH_9rXfTYeLLeLr:bgd9k5 -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/spanset_test.go:
```diff
@@ -394,3 +394,237 @@ func TestSpanSetWriteImpliesRead(t *testing.T) {
+		expected bool
+	}{
+		{
+			name:     "s1 contains s2 exactly",
```
Done.
<!-- Reviewable comment -OeNsu-A1w32t_F5AKf4:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157153)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -382,24 +387,63 @@ func (s *SpanSet) checkAllowed(
-		return s1.Contains(s2)
+// Contains returns whether s1 contains s2, where s2 can be a TrickySpan.
+func Contains(s1 roachpb.Span, s2 TrickySpan) bool {
+	s2Span := roachpb.Span(s2)
```
nit: defer this conversion as far down the stack as possible and best avoid it. The idea is that the type carries semantics, so we don't want to lose/disperse it in the first line.

```golang
if s2.Key != nil { // TrickySpan is a normal roachpb.Span
	return s1.Contains(roachpb.Span(s2))
}

// below, it's only Key/EndKey arithmetics, so we don't
// need s2 to be roachpb.Span.
```

iskettaneh replied:
Done! With the new helper function, I think you are going to like how Contains and Overlaps look now!
<!-- Reviewable comment -OeScB_I-sDR1Era6ZCm:b-nbq9id -->
<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -353,7 +358,7 @@ func (s *SpanSet) checkAllowed(
 		// Check if the span is forbidden.
 		for _, matcher := range s.forbiddenSpansMatchers {
-			if err := matcher(span); err != nil {
+			if err := matcher(TrickySpan(span)); err != nil {
```
I think we should make the `span` in this method signature a `TrickySpan`. More generally, do this all the way up the stack to where the span is created. Specifically, those places ([example](https://github.com/cockroachdb/cockroach/blob/560b635c16f790db75a0fdc7492be9ecd8e96fa6/pkg/kv/kvserver/spanset/batch.go#L284)) where it is created with a nil `Key` will now start making sense, because the `TrickySpan` type is a hint.

iskettaneh replied:
Done.
<!-- Reviewable comment -OeSgZOq82NbR91Dlbk5:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Done.
<!-- Reviewable comment -OeSgZOq82NbR91Dlbk5:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Done.
<!-- Reviewable comment -OeSgZOq82NbR91Dlbk5:b-896fix -->
<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
This is nearing the LGTM, thanks for making it cleaner each time.

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,76 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+
+	// If we are partially overlapping the full rangeID local span. We must be
+	// overlapping rangeID local keys.
+	if spanset.PartiallyOverlaps(fullRangeIDLocalSpans, span) {
```
Good point!
<!-- Reviewable comment -OeRf1cN6b4Ig8ua-bGt:bdcv6kv -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Good point!
<!-- Reviewable comment -OeRf1cN6b4Ig8ua-bGt:bdcv6kv -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -382,24 +387,63 @@ func (s *SpanSet) checkAllowed(
-	return s1.Key.Compare(s2.EndKey) < 0 && s1.EndKey.Compare(s2.EndKey) >= 0
+	// s1 is [s1.Key, s1.EndKey), s2 is [s2.EndKey.Prev(), s2.EndKey)
+	// They overlap iff s2.EndKey.Prev() is in [s1.Key, s1.EndKey).
+	return s1.Key.Compare(s2Span.EndKey) < 0 && s1.EndKey.Compare(s2Span.EndKey) >= 0
```
Done.
<!-- Reviewable comment -OeSN7Ma38LZOa-F2Yfa:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Done.
<!-- Reviewable comment -OeSN7Ma38LZOa-F2Yfa:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157153)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -76,6 +76,11 @@ type Span struct {
+// TrickySpan represents a span that supports a special encoding where a nil
+// start key with a non-nil end key represents the point span:
+// [EndKey.Prev(), EndKey).
+type TrickySpan roachpb.Span
```
Yeah I don't think we can unexport it.
<!-- Reviewable comment -OeSdiJpANESsRydHIke:b-jdiryj -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Yeah I don't think we can unexport it.
<!-- Reviewable comment -OeSdiJpANESsRydHIke:b-jdiryj -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,76 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+// with any unreplicated rangeID local keys.
+// Note that we could receive the span with a nil startKey, which has a special
+// meaning that the span represents: [endKey.Prev(), endKey).
+func overlapsUnreplicatedRangeIDLocalKeys(span spanset.TrickySpan) error {
```
Thank you!
<!-- Reviewable comment -OeRf4RZ4hJNY0JrFd7X:b-9755fn -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Thank you!
<!-- Reviewable comment -OeRf4RZ4hJNY0JrFd7X:b-9755fn -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,76 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	getRangeIDLocalSpan := func(rangeID roachpb.RangeID) roachpb.Span {
+		return roachpb.Span{Key: keys.MakeRangeIDUnreplicatedPrefix(rangeID),
+			EndKey: keys.MakeRangeIDUnreplicatedPrefix(rangeID).PrefixEnd()}
+	}
```
Done.
<!-- Reviewable comment -OeRfJVy05kTRms5C7gK:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Done.
<!-- Reviewable comment -OeRfJVy05kTRms5C7gK:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Done.
<!-- Reviewable comment -OeRfJVy05kTRms5C7gK:b-896fix -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,76 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	// Check to see if we fully contain fullRangeIDLocalSpans.
+	if !spanset.Contains(fullRangeIDLocalSpans, span) {
+		return errors.Errorf("completely containing the unreplicated rangeID span")
+	}
```
Can you take a look at the function now? It has less checks and I think it's decently readable.
<!-- Reviewable comment -OeS3OpNAAEXMHTryQLU:b-hoytla -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Can you take a look at the function now? It has less checks and I think it's decently readable.
<!-- Reviewable comment -OeS3OpNAAEXMHTryQLU:b-hoytla -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
Can you take a look at the function now? It has less checks and I think it's decently readable.
<!-- Reviewable comment -OeS3OpNAAEXMHTryQLU:b-hoytla -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -888,6 +888,59 @@ func DisableReadWriterAssertions(rw storage.ReadWriter) storage.ReadWriter {
+		return &spanSetBatch{
+			ReadWriter: ReadWriter{
+				spanSetReader: spanSetReader{r: v.b, spans: spans, spansOnly: v.spansOnly, ts: v.ts},
+				spanSetWriter: spanSetWriter{w: v.b, spans: spans, spansOnly: v.spansOnly, ts: v.ts},
```
I refactored the code a bit and I introduced SpanSet ShallowCopy which helped clean things up.
<!-- Reviewable comment -OeTG8U3BzY53Rezm7Qj:b-2fala3 -->
<!-- Sent from Reviewable.io -->

iskettaneh replied:
I refactored the code a bit and I introduced SpanSet ShallowCopy which helped clean things up.
<!-- Reviewable comment -OeTG8U3BzY53Rezm7Qj:b-2fala3 -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -888,6 +888,59 @@ func DisableReadWriterAssertions(rw storage.ReadWriter) storage.ReadWriter {
+			},
+			b:         v.b,
+			spans:     spans,
+			spansOnly: v.spansOnly,
```
I added a TODO to address that. I think some of these fields aren't even read.
<!-- Reviewable comment -OeTGq3lBFP_F2PxbcwY:b1kfbfy -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -888,6 +888,59 @@ func DisableReadWriterAssertions(rw storage.ReadWriter) storage.ReadWriter {
+			spans:                  v.ReadWriter.spanSetReader.spans.spans,
+			forbiddenSpansMatchers: v.ReadWriter.spanSetReader.spans.forbiddenSpansMatchers,
+			allowForbidden:         v.ReadWriter.spanSetReader.spans.allowForbidden,
+			allowUndeclared:        v.ReadWriter.spanSetReader.spans.allowUndeclared,
```
With the new refactoring, I believe things are clearer now.
<!-- Reviewable comment -OeTGy9NAYILVpUfK7A1:bgg4nw2 -->
<!-- Sent from Reviewable.io -->

iskettaneh commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,76 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+		return errors.Errorf("completely containing the unreplicated rangeID span")
+	}
+
+	// We make an assumption here that whoever if the span in inside
```
I changed it
<!-- Reviewable comment -OeRjt_7BicolGXGn3D1:b-k6aiwd -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157153)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157153)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -201,14 +221,22 @@ func (s *SpanSet) AddMVCC(access SpanAccess, span roachpb.Span, timestamp hlc.Ti
 	}
+	s.forbiddenSpansMatchers = append(s.forbiddenSpansMatchers, s2.forbiddenSpansMatchers...)
 	s.allowUndeclared = s2.allowUndeclared
+	s.allowForbidden = s2.allowForbidden
```
pre-existing: the logic is a bit unexpected. Maybe it should be

```golang
s.allowUndeclared = s.allowUndeclared || s2.allowUndeclared
// same for allowForbidden
```

A TODO to figure that out would be good. Yet another reason to segregate the data structure from assertions.

pav-kv commented on pkg/kv/kvserver/spanset/spanset.go:
```diff
@@ -392,16 +397,11 @@ func (s *SpanSet) checkAllowed(
-	}
-
+// doesNormalSpanContainPointTrickySpan takes a normal span (s1) that has
+// both Key and EndKey not nil, and takes a tricky span s2, where the Key is
```
`s1.EndKey` can be nil. Can just say that it's a normal `Span`.

pav-kv commented on pkg/kv/kvserver/spanset/spanset_test.go:
```diff
@@ -203,25 +204,25 @@ func TestSpanSetCheckAllowedAtTimestamps(t *testing.T) {
 	ss.AddNonMVCC(SpanReadWrite, roachpb.Span{Key: keys.RangeGCThresholdKey(1)})
 
 	var allowedRO = []struct {
```
No tests here with `TrickySpan{EndKey: something}`. Could be good to add some. We have good coverage for it, so optional / TODO.

More generally, would be also good to de-boilerplate these tests. Also feel free to ignore / TODO.

Same below.

pav-kv commented on pkg/kv/kvserver/spanset/spanset_test.go:
```diff
@@ -381,16 +382,132 @@ func TestSpanSetWriteImpliesRead(t *testing.T) {
+		expected bool
+	}{
+		// s1 equals s2.
+		{s1: "a-c", s2: "a-c", expected: true},
```
I think in situations like this it's justifiable. As a reviewer, I'd be happier to review one list rather than two. Plus, the benefit that we can check relations between the two funcs, as an extra assertion.

This relates to one of my previous points (in this instance an overkill, but consider this logically): these 2 funcs could be one - `DiffSpans(s1, s2) -> descriptor` where the "descriptor" tells us everything about the relative position of the two spans. From it, we can say whether they overlap, contained, etc. So logically it's one test: for 2 spans - what "descriptor" do we expect.

pav-kv commented on pkg/kv/kvserver/batcheval/cmd_end_transaction.go:
```diff
@@ -1340,8 +1340,10 @@ func splitTriggerHelper(
+
 	if err := storage.MVCCPutProto(
-		ctx, batch, keys.RangeLastReplicaGCTimestampKey(split.RightDesc.RangeID), hlc.Timestamp{},
+		ctx, spanset.DisableForbiddenSpanAssertionsOnBatch(batch),
```
In `DisableForbiddenSpanAssertionsOnBatch` definition, pop in a TODO to eliminate all its users?

pav-kv commented on pkg/keys/constants.go:
```diff
@@ -163,6 +163,7 @@ var (
 	//
 	// LocalStorePrefix is the prefix identifying per-store data.
 	LocalStorePrefix = makeKey(LocalPrefix, roachpb.Key("s"))
+	LocalStoreMax    = roachpb.Key(LocalStorePrefix).PrefixEnd()
```
nit: `LocalStorePrefixEnd`

pav-kv commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -15197,3 +15197,110 @@ func TestLeaderlessWatcherInit(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testCases := []struct {
+		name      string
```
unused

pav-kv reviewed (commented):
The non-test parts LGTM, minor suggestions. Request to clean up the tests a bit, so that I can make a final (self-convincing) pass on them and LGTM tomorrow.

pav-kv commented on pkg/kv/kvserver/spanset/spanset_test.go:
```diff
@@ -381,16 +388,106 @@ func TestSpanSetWriteImpliesRead(t *testing.T) {
+
+	if parts[1] != "X" {
+		end = roachpb.Key(parts[1])
+	}
```
style nit (feel free to ignore, just noticed the pattern in this PR):

In simple code, vertical sparseness / empty lines is often not needed. Empty lines is a tool to communicate/separate slightly more medium-size blocks of logically related code.

E.g. here I would recommend making it one block:

```golang
var start, end roachpb.Key
if parts[0] != "X" {
	start = roachpb.Key(parts[0])
}
if parts[1] != "X" {
	end = roachpb.Key(parts[1])
}
```

Also, could make it more concise like:

```golang
var span roachpb.Span
if parts[0] != "X" {
	span.Key = roachpb.Key(parts[0])
}
if parts[1] != "X" {
	span.EndKey = roachpb.Key(parts[1])
}
return span
```

pav-kv commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -884,3 +889,66 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+
+	// At this point, we know that we overlap with fullRangeIDLocalSpans. If we
+	// are not completely within fullRangeIDLocalSpans, return an error as we
+	// collide with at least one unreplicated RangeIDLocal key.
```
Technically, it's possible that we don't collide. E.g. you can imagine a little span that begins before the RangeID-local space, and slightly moves into it without touching the first range's unreplicated.

We just make it a "policy" here that it shouldn't be possible, i.e. the spans are sensibly respecting the "tree" structure, and a span that begins outside the subtree and ends inside (or vice versa) doesn't make logical sense.

Dunno it there is concise way to say this, see if you can compress it :)

pav-kv commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -15197,3 +15197,123 @@ func TestLeaderlessWatcherInit(t *testing.T) {
+// TestOverlapsStoreLocalKeys verifies that the function
+// overlapsStoreLocalKeys() successfully catches any overlap with
+// store local keys.
+func TestOverlapsStoreLocalKeys(t *testing.T) {
```
Nice, thanks a lot for the clean up! Was able to review it in a minute.

pav-kv commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -15197,3 +15197,123 @@ func TestLeaderlessWatcherInit(t *testing.T) {
+		// Tricky spans not overlapping with unreplicated local RangeID spans.
+		{span: s(nil, keys.LocalRangeIDPrefix.AsRawKey())},
+		{span: s(nil, keys.RangeForceFlushKey(1))},
+		{span: s(nil, keys.LocalRangeIDPrefix.AsRawKey().PrefixEnd().Next())},
```
Can add one without `Next()`, it should be legit too (IIUC it will take a different branch in the func).

pav-kv commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -15197,3 +15197,123 @@ func TestLeaderlessWatcherInit(t *testing.T) {
+		{span: s(keys.RaftTruncatedStateKey(1), nil), notOk: true},
+		{span: s(keys.RaftTruncatedStateKey(2), nil), notOk: true},
+
+		// Tricky spans not overlapping with unreplicated local RangeID spans.
```
A couple of corner cases around the unreplicated span edges.

pav-kv commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -15197,3 +15197,123 @@ func TestLeaderlessWatcherInit(t *testing.T) {
+		{span: s(keys.RaftTruncatedStateKey(1), keys.RaftTruncatedStateKey(2)), notOk: true},
+		{span: s(keys.RaftTruncatedStateKey(1), roachpb.KeyMax), notOk: true},
+
+		// Point spans not overlapping with unreplicated local RangeID spans.
```
A couple of corner cases around the unreplicated span edges.

Applies to a few blocks here. Could be convenient to pre-compute a couple of `var`s for some unreplicated span and use it in a few places:

```golang
r1Unrepl := roachpb.Span{...}

{span: s(r1Unrepl.Key.Prev(), nil)},
{span: s(r1Unrepl.EndKey, nil)},

// further down
{span: s(nil, r1Unrepl.EndKey), notOk: true},
// etc
```

pav-kv reviewed (approved):
LGTM, thanks for this change! Some last suggestions included, apply as many as makes sense.
